Launch Security with AMD SEV-ES

Fuzzy-Math opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**: 

This PR extends the functionality of AMD SEV in Kubevirt by enabling the SEV-ES extension. Under SEV-ES, both the VM's memory and the VM's register state are encrypted.

The VMI spec is extended to support setting the EncryptedState SEV policy in order to launch a VM with SEV-ES features enabled.

Please refer to #6166 for more information and known limitations. General information on AMD SEV-ES can be found at [AMD's developer portal](https://developer.amd.com/sev/)

**Special notes for your reviewer**:

- `WorkloadEncryptionSEV` feature gate needs to be enabled
- A node with an AMD EPYC CPU with support for SEV-ES is needed in the cluster
- The following should be added to the VMI spec to request SEV-ES launch security:
```yaml
spec:
  domain:
    launchSecurity:
      sev:
        policy:
          encryptedState: true

```
**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Experimental support for AMD SEV-ES
```

vasiliy-ul commented on pkg/virt-handler/node-labeller/cpu_plugin.go:
```diff
@@ -189,6 +189,15 @@ func (n *NodeLabeller) getDomCapabilities() (HostDomCapabilities, error) {
+	}
+
+	if hostDomCapabilities.SEV.Supported == "yes" && hostDomCapabilities.SEV.MaxESGuests > 0 {
+		hostDomCapabilities.SEV.SupportedES = "yes"
```
`SupportedES = "yes"` seems redundant since you already have `MaxESGuests > 0`. What is the reason of introducing `SupportedES`?

vasiliy-ul commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -127,19 +127,25 @@ var _ = Describe("Node-labeller ", func() {
 	It("should add host cpu model label", func() {
-		expectNodePatch(kubevirtv1.HostModelCPULabel)
 		res := nlController.execute()
+		expectNodePatch(kubevirtv1.HostModelCPULabel)
```
why would you need to change this test (and the one bellow)? Also it does not seem correct to setup the expectations **after** calling `execute()` (this applies to all the tests here).

vasiliy-ul commented on tests/launchsecurity/sev.go:
```diff
@@ -19,23 +19,33 @@ var _ = Describe("[sig-compute]AMD Secure Encrypted Virtualization (SEV)", func(
+		func(isESEnabled bool) {
+			const secureBoot = false
+			dmesgRet := "SEV"
+			if isESEnabled {
```
It would be bettter to skip the test if there is no ES-capable node in the cluster. You can refer to the implementation of the function `checks.SkipTestIfNotSEVCapable()` (which is called here in `BeforeEach`) and implement a similar one for ES:

https://github.com/kubevirt/kubevirt/blob/409494e0c5b34f4a1cb5b251954eee3404654c67/tests/framework/checks/skips.go#L117

vasiliy-ul reviewed (commented):
Looks good :+1: . I left a couple of remarks inline.

Also I would suggest to folllow [imperative mood](https://cbea.ms/git-commit/#imperative) in the commits subject line. E.g. `Adding ...` -> `Add ...`, `Generated ...` -> `Generate ...`.

iholder-redhat commented on tests/libvmi/vmi.go:
```diff
@@ -117,10 +117,14 @@ func WithUefi(secureBoot bool) Option {
 
 // WithSEV adds `launchSecurity` with `sev`.
-func WithSEV() Option {
+func WithSEV(isESEnabled bool) Option {
```
What happens in the future when we'll also support SNP? Would it make sense to have a following function then? Or SNP has to be defined with ES?
```go
func WithSEV(isESEnabled, isSNPEnabled bool) Option {
```

Maybe it's better to define an explicit type to choose the different policies:
```go
type AmdSevPolicy string
const(
  Sev = "sev"
  SevES = "sev-es"
  SevSNP = "sev-snp"
)

func WithSEV(sevPolicy AmdSevPolicy) Option { ...
```

Fuzzy-Math replied:
I won't reason much about SNP since we don't know how libvirt will expose support for it. If it's exposed as a policy bit like SEV-ES, then I think the first approach is better since it will set the vmi spec directly. For SEV & SEV-ES, I think what is there currently is the cleanest approach.

iholder-redhat reviewed (commented):
Great Job @Fuzzy-Math! left some comments.
By the way, does ES affect the attestation process in any way?

vasiliy-ul commented on pkg/virt-launcher/virtwrap/launchsecurity/sev.go:
```diff
@@ -0,0 +1,43 @@
+
+func SEVPolicyToBits(policy *v1.SEVPolicy) uint {
+	// NoDebug is always true
+	bits := uint(SEVPolicyNoDebug)
```
`NoDebug` is always set to true just because of security reasons. If we do not explicitly disable debugging, the hypervisor will be able to decrypt and read the guest memory via the `DBG_DECRYPT/DBG_ENCRYPT` API. Something that we likely would prefer to avoid.

vasiliy-ul replied:
I would say rather no, because its a big security hole that makes the whole idea of confidential VMs useless (potentially anyone with access to the hypervisor will be able to read sensitive data.... and that is what SEV primaraly tries to address).

iholder-redhat replied:
I tend to agree with @vasiliy-ul. We can always add such capability if someone needs it (and probably warn the user that of the risks involved). For now I think we can stick with the safer approach.

Thanks for the clarifications!

Fuzzy-Math commented on pkg/virt-handler/node-labeller/cpu_plugin_test.go:
```diff
@@ -163,24 +163,30 @@ var _ = Describe("Node-labeller config", func() {
 
 	Context("return correct SEV capabilities", func() {
-		It("when SEV is supported", func() {
+		It("when SEV(ES) is supported", func() {
```
This was something I tried during development. We actually have 3 cases to test: SEV and SEV-ES, only SEV, and neither SEV nor SEV-ES, of which only the first and last are tested here. That third case involves keeping a separate xml (domcapabilities_sev_es.xml) that only differs by one line to the existing xml, I don't feel that's worth it. I also tried writing to the xml file and rewriting it back in that test case, but that wasn't a good solution either. Is there a way to patch a single xml node without having to touch the rest? Should we just introduce another xml testdata file for sev-es even if only one line differs?

vasiliy-ul commented on pkg/virt-handler/node-labeller/model.go:
```diff
@@ -92,6 +92,9 @@ type Feature struct {
-	ReducedPhysBits string `xml:"reducedPhysBits"`
+	CBitPos         uint   `xml:"cbitpos"`
+	ReducedPhysBits uint   `xml:"reducedPhysBits"`
+	MaxGuests       uint   `xml:"maxGuests"`
+	MaxESGuests     uint   `xml:"maxESGuests"`
```
If the limits are exceeded, the firmware will just return an error and qemu will fail to start. Ideally we need to consider those limits in the device controller when we setup the plugin for `/dev/sev`. Currently the plugin advertizes the default value (which is the same as for `/dev/kvm` and friends, i.e. `1k`):

https://github.com/kubevirt/kubevirt/blob/3da05ab734d318c7fc3e99fabbb24f5e5de95c4d/pkg/virt-handler/device-manager/device_controller.go#L103-L110

So, answering the initial question: the VM will be scheduled but it will fail to start. I would say that it is smth to fix in a follow-up PR :)

> Does AMD document these limits for specific CPU generations or firmware versions?

I am pretty sure it does. Though I think we are mostly interested in getting those during runtime.

> Does the entire stack keep the counts up to date as VMs are scheduled on a node? What would it look like for Kubevirt to query this each time we try to launch an SEV/SEV-ES guest? If the count isn't reliable, can kubevirt keep internal counts of the number of SEV/SEV-ES guests?

I think there is no need for KubeVirt to query it each time (and I dont think there is an API for getting the current ammount of SEV guests). As mentioned above I would rather offload the accounting to the kuberbetes device mgmt.

iholder-redhat replied:
Thanks for the clarifications!

Indeed this sounds like something to tackle in a follow-up PR.

At a first glance it makes sense to me that virt-handler would have an internal counter and would mark the node with a label if no more SEV workloads can be scheduled.

vasiliy-ul commented on pkg/virt-handler/node-labeller/model.go:
```diff
@@ -92,6 +92,9 @@ type Feature struct {
+	ReducedPhysBits uint   `xml:"reducedPhysBits"`
+	MaxGuests       uint   `xml:"maxGuests"`
+	MaxESGuests     uint   `xml:"maxESGuests"`
+	SupportedES     string `xml:"-"`
```
Just as some historical background behind the `Supported` field in this struct: the previous libvirt version did not advertise `MaxGuests`. Therefore `supported='yes'` was the only way to figure out if SEV was supported on the host. 

Also if SEV is not in supported then the xml will look smth like bellow:
```xml
    <sev supported='no' />
```
And the other struct fields will be initialized to their defaults (`0` for ints and `""` for strings). So it is the only logical way to query for SEV.

Fuzzy-Math replied:
What are your thoughts with trying to upstream a `supportedES` attribute into libvirt and make that the reliable way to query ES support? Then let max(ES)Guests be used for scheduling logic as discussed below.

vasiliy-ul replied:
honestly, I dont see what makes `maxESGuests` 'not reliable'.

iholder-redhat replied:
Thanks for all the interesting information!

I think that "not reliable" should be replaced with "not obvious". If it's possible, I would push it to libvirt's upstream.

BTW, if `xml:"-"` simply means to ignore it when marshalling / unmarshalling - isn't it the same as not having this tag at all?

vasiliy-ul commented on pkg/util/util.go:
```diff
@@ -83,6 +83,14 @@ func IsSEVVMI(vmi *v1.VirtualMachineInstance) bool {
 	return vmi.Spec.Domain.LaunchSecurity != nil && vmi.Spec.Domain.LaunchSecurity.SEV != nil
 }
 
+// Check if a VMI spec requests AMD SEV
```
```suggestion
// Check if a VMI spec requests AMD SEV-ES
```

vasiliy-ul reviewed (commented):
/ok-to-test

alicefr commented on pkg/virt-handler/node-labeller/cpu_plugin.go:
```diff
@@ -189,6 +189,15 @@ func (n *NodeLabeller) getDomCapabilities() (HostDomCapabilities, error) {
+		return hostDomCapabilities, err
+	}
+
+	if hostDomCapabilities.SEV.Supported == "yes" && hostDomCapabilities.SEV.MaxESGuests > 0 {
```
Is SEV struct always set or can it be nil?

Fuzzy-Math replied:
I tested this by removing the entire sev node from `pkg/virt-handler/node-labeller/testdata/domcapabilities_sev.xml` and the SEV struct is still not nil.

alicefr commented on pkg/util/util.go:
```diff
@@ -83,6 +83,14 @@ func IsSEVVMI(vmi *v1.VirtualMachineInstance) bool {
 }
 
+// Check if a VMI spec requests AMD SEV
+func IsSEVESVMI(vmi *v1.VirtualMachineInstance) bool {
```
Can we add the check if the vmi is equal to nil and return false if so?

Fuzzy-Math replied:
Everything else in the util package assumes the vmi instance is present and valid or is handled in the caller before calling into the util package. I think it's safe to assume the same for SEV and SEV-ES and not add further clutter here.

alicefr commented:
I left a couple of questions but looks good to me :)

Fuzzy-Math commented:
/remove-lifecycle rotten

Fuzzy-Math commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

Fuzzy-Math commented:
/hold until issues with functional test are resolved

Fuzzy-Math commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

Fuzzy-Math commented:
Updates to this PR:
- A new domcapabilities_noseves.xml testdata file was added to support all 3 sev cases for node labeller (just sev, sev and sev-es, neither sev nor sev-es)
- A previous commit in this series modified the usage of the sev node label. Since the cpu realtime label was also modified and because there were some further issues I found with those modifications, I have moved those changes to another PR (#9430)

Now that hardware has been added to the CI cluster to accomodate SEV-ES, this PR is now ready for approval.

Fuzzy-Math commented:
/unhold

Fuzzy-Math commented:
/hold again as we are blocked #9430. We rely on the PR's implementation of addNodeSelector() inside vmi-mutator.go

Fuzzy-Math commented:
/unhold, this PR has gone back to using empty node labels for SEV/SEV-ES; #9430 will address empty labels independently.

Fuzzy-Math commented:
/unhold

Fuzzy-Math commented:
/test pull-kubevirt-e2e-k8s-1.26-sev

Fuzzy-Math commented:
@iholder101 @alicefr @xpivarc @vasiliy-ul just a reminder that this PR can be reviewed independently of other SEV-related PRs and is ready to progress.

> Updates to this PR:
> 
> * A new domcapabilities_noseves.xml testdata file was added to support all 3 sev cases for node labeller (just sev, sev and sev-es, neither sev nor sev-es)
> * A previous commit in this series modified the usage of the sev node label. Since the cpu realtime label was also modified and because there were some further issues I found with those modifications, I have moved those changes to another PR ([Ensure node label values are not empty #9430](https://github.com/kubevirt/kubevirt/pull/9430))
> 
> Now that hardware has been added to the CI cluster to accomodate SEV-ES, this PR is now ready for approval.

vasiliy-ul reviewed (approved):
From my perspective, this PR is good to get in. Most of the comments have been addressed and the tests on SEV-ES hardware are green.

vasiliy-ul commented:
Hey, @xpivarc, @alicefr, @iholder101, could you take one more look here? All the previous comments seem addressed, and the CI is happy. Can we move this forward?

alicefr commented:
@vasiliy-ul yes, we should definitely move this forward. I'll review it tomorrow

alicefr commented:
/approve
/hold

Let's give some time to @xpivarc and @iholder101 to have a second look

vasiliy-ul commented:
I would like to unhold the PR and move it forward. Discussed and agreed on that with @xpivarc.

@iholder101, you previously reviewed the code here. Any objections? Or do you want to take another look? The code has not changed significantly since then, as far as I can see.

vasiliy-ul commented:
/lgtm
/unhold
/test pull-kubevirt-e2e-k8s-1.26-sev

vasiliy-ul commented:
/retest-required

vasiliy-ul commented:
/test pull-kubevirt-e2e-k8s-1.25-sig-storage

Fuzzy-Math commented:
/retest-required
